File: 24. Swap Nodes in Pairs/SwapNodesInPairs.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	var resultHead *ListNode
	var resultTail *ListNode
	traversingNode := head
	for traversingNode != nil {
		//create swap pair
		swappedNode := ListNode{Val: traversingNode.Val}
		if traversingNode.Next != nil {
			swappedNode = ListNode{
				Val:  traversingNode.Next.Val,
				Next: &ListNode{Val: traversingNode.Val},
			}
		}

		//attach swapped pair
		if resultHead == nil {
			resultHead = &swappedNode
			resultTail = resultHead
		} else {
			resultTail.Next = &swappedNode
			resultTail = resultHead.Next
		}
		if resultTail.Next != nil {
			resultTail = resultTail.Next
		}

		//go to next element
		traversingNode = traversingNode.Next
		if traversingNode != nil {
			traversingNode = traversingNode.Next
		}
	}

	return resultHead
}

func main() {
	head := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: &ListNode{
						Val: 5,
					},
				},
			},
		},
	}

	reversed := swapPairs(head)
	for reversed != nil {
		fmt.Println(reversed.Val)
		reversed = reversed.Next
	}
}

File: 24. Swap Nodes in Pairs/SwapNodesInPairs.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        resultHead: Optional[ListNode] = None
        resultTail: Optional[ListNode] = None
        traverseNode = head
        while(traverseNode is not None):
            next = traverseNode.next
            elementsToAdd = ListNode(traverseNode.val) if next is None else ListNode(next.val, ListNode(traverseNode.val))
            if resultHead is None:
                resultHead = elementsToAdd
                resultTail = resultHead if resultHead.next is None else resultHead.next
            else:
                resultTail.next = elementsToAdd
                resultTail = resultTail.next if resultTail.next is None else resultTail.next.next
            traverseNode = traverseNode.next if traverseNode.next is None else traverseNode.next.next

        return resultHead
    
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
solution = Solution()
traversedHead = solution.swapPairs(head)
while traversedHead is not None:
    print(traversedHead.val)
    traversedHead = traversedHead.next